design/status-design-update.md

```diff
@@ -0,0 +1,1166 @@
+# HTTPProxy status redesign
+
+## Background
+
+Given its complexity, Contour's HTTPProxy CRD currently has insufficient information about what its status surfaced to the user.
+
+Currently, it only supports a `currentStatus` and `description`, which are particularly poor user experience around inclusion and conditions.
+
+In particular, the following questions about HTTPProxy are particularly problematic at the moment:
+
+- What happens when *some* parts of a HTTPProxy configuration are invalid? What if one include is invalid, and the other is not? How can the user be aware of what's happening?
+- What happens if there are problems on inclusion, particularly condition conflicts?
+- What happens if there is a problem with composable HTTPProxy use, that is, including the same HTTPProxy object from multiple parents.
+- How do users understanding the full tree of includes used to build an Envoy config from multiple HTTPProxy objects?
+
+## Design Goals
+
+### Goals Overview
+
+In rethinking Status, we need to ensure that the whatever we choose can fit all the use cases our HTTPProxy design allows for.
+
+In particular, the features that strongly affect how we design status:
+
+- Inclusion: We need to have a clear set of rules about what classes of errors show up at the *includer* level and what at the *includee* level.
+  - We also don't put limits on how big this structure should be, so we need to ensure that the status block can provide useful information for large levels of complexity.
+- Composability: Because HTTPProxies can be included under more than one root, and can also be included in multiple places under a single root, Status needs to be able to work in both those cases.
+
+We also need to consider the interaction between service admins when one side changes something that breaks something for the other side.
+
+In addition, we need to keep some rules in mind about making the status useful:
+
+- the status must be able to indicate multiple errors, not just one.
+- errors must be able to be directly attributed to the part of the config that produced them.
+  - To help with this, the status should have a way to show errors with routes and includes separately.
+  - Most portions of the status must be namespaced by at least the fqdn of the root, if not some rendering of the include path.
```

Comment:
Namespaced as in Kubernetes namespaces? We need to be careful not to leak information to users that might not be privileged.

Reply:
Ah, term overloading. No, not namespaced as in Kubernetes namespaces, just by the FQDN of the root HTTPProxy. I think it's okay to leak where your includers are coming from, since you should have RBAC stopping you from interacting with them.